### ISSUE TITLE
Pinning messages

### ISSUE THREAD

**Issue by jakobroehrl** [nextcloud/spreed#3390]
In a big room it's sometimes hard to find important messages.
Maybe it would be nice to favourite some of them and have the possibility to show only favourited messages.

What do you think about this?

**Comment by nickvergessen** [comment]
Favorites not so much, but I couldimagine to ping some messages into a chat which then are shown always on top, similar to the featue in Telegram and Line

**Comment by Nyco** [comment]
Favorites and pins usually do not work in terms of UX and usage.

**Comment by jancborchardt** [comment]
@Nyco is there some user research about that? Whenever doing a :-1: on a comment please explain it more, otherwise it comes off as a bit unfriendly. ;)

Agree with @nickvergessen, others offer this, Slack as well. To be on par with expectations, this is something we should have at some point.

**Comment by Nyco** [comment]
> Whenever doing a 👎 on a comment please explain it more, otherwise it comes off as a bit unfriendly. ;)

Oh sorry, I apologise.
I am a bit confused, I thought thumbs up and down were simples votes.

> @Nyco is there some user research about that?

Yes, we had interviewed people on the use of personal chats (Telegram, WhatsApp, Messenger,...) and team chats (Slack, Discord, Teams...), and it turns out people just don't use pins and favorites for messages: either they just don't see (anymore) or filter out that feature, or they just don't see any need. Sometimes they also find the UI to be polluting.

**Comment by nickvergessen** [comment]
Nearly all my Telegram and Line group chats have a pinned message.
I guess it's just a matter of habbits.

**Comment by jancborchardt** [comment]
Yeah, first off it would be good to know _how many_ people were interviewed, since that could be anecdotal or biased. I would also take interviews with a grain of salt in general – asking people what they do or want generally gives you a less accurate answer than just watching what they actually do. (Please do clarify if I’m misunderstanding what you mean by "interviews".)

**Comment by HammyHavoc** [comment]
Being able to favourite messages and link to them directly is pretty important and one of the major features of Rocket.Chat that we're sorely missing.

**Comment by PVince81** [comment]
> Being able to favourite messages and link to them directly is pretty important and one of the major features of Rocket.Chat that we're sorely missing.

Direct link to messages is technically already possible.
When searching for a message it generates a URL that makes the list scroll to the target message.
What's missing is the ability to copy that message link directly from the message row.

**Comment by PVince81** [comment]
I've raised https://github.com/nextcloud/spreed/issues/5097 for copying the direct link

**Comment by jospoortvliet** [comment]
I think there's pinning and favoriting.

❤️  Favoriting should perhaps show the messages in the Media tab as the 3rd tab there - the use case would be resources shared in a chat at some point. Esp in a big group chat, important messages get lost quickly, and it'd be nice if you could favorite them. It'd be individual, just for your own 'documentation'.

📌  Pinning is more of a global thing - a message gets pinned on top by a moderator, so everyone sees it. Mostly used for things like announcements.

For favoriting I guess we need a new issue?

**Comment by dinosmm** [comment]
Message pinning is even more important now I just realised that the Projects functionality was unexpectedly dropped. 
Now there is no way to have a permanent read/write folder/file attached and always shown in a Talk chat. 

Please do bring some form of message pinning, it is important for many use cases, no less for participants entering a chat can quickly find and open a collaborative document that they have to be working on.

**Comment by nickvergessen** [comment]
> Message pinning is even more important now I just realised that the Projects functionality was unexpectedly dropped.

you can still reenable it with a config/config.php:
```
'projects.enabled' => true,
```

> Now there is no way to have a permanent read/write folder/file attached and always shown in a Talk chat.

We use the description for that, works perfectly fine as well:
![grafik](https://user-images.githubusercontent.com/213943/233580046-083539d8-dad9-4257-aaaa-6d9b138e105d.png)

**Comment by dinosmm** [comment]
> you can still reenable it with a config/config.php:

Ah thanks, good to know. However, is this going to be fully deprecated and removed in a future version? No good getting used to using it if it's going to go away completely in the next year or two.

> We use the description for that, works perfectly fine as well:

Good idea, that actually works better than the Project function, as guests that don't have an NC account don't actually see the shared items tab.

**Comment by WilliamDEdwards** [comment]
The ability to pin messages in high-traffic chats would be very useful.

**Comment by andynameistaken** [comment]
Organisation that I am providing Nextcloud to is critically needing it too. It has ~300 users and not being able to pin the message on top of the chat decreases usability of Talk big time. People were used to it from Discord which I made migration from and there was even proposition of going back to the Discord because of lack of this feature, which makes me sad because Discord is not Open Source. 
Date of creation of this issue doesn't give me high hopes tho.

**Comment by WilliamDEdwards** [comment]
> Organisation that I am providing Nextcloud to is critically needing it too. It has ~300 users and not being able to pin the message on top of the chat decreases usability of Talk big time. People were used to it from Discord which I made migration from and there was even proposition of going back to the Discord because of lack of this feature, which makes me sad because Discord is not Open Source. Date of creation of this issue doesn't give me high hopes tho.

Solution: make a PR :)

**Comment by miaulalala** [comment]
Could we add the pinned messages somewhere in the right sidebar?

Shitty mockup:
![image](https://github.com/user-attachments/assets/e0dd85a7-2fe2-40e9-8f9a-4086a5357b47)

**Comment by Antreesy** [comment]
> in the right sidebar?

I'd say, it would make sense only if we allow to pin several messages at once (IIRC Telegram allows it, but just navigating through the chat from recent to older pinned message).
Though that might quickly rise to the ton of chat functionality duplication here, unless we restrict it to simplified view like on your mockup

**Comment by jancborchardt** [comment]
The intention of pinning a message is to highlight it especially, in semi-permanent way, almost like the description but not quite. So:
- It should be very presently visible (sidebar is not so ideal)
- It is somewhat temporary and immediately relevant info. Anything relevant for longer can go into the description
- It should be a single focus at a time, so only 1 pinned message (new pinned message replaces old one, people of course need to be informed about that if one exists already in the conversation)


Quick inspector mockup:
- The message is pinned as is, floating on the top below the heading
- Shows who pinned it behind the name
- Ellipsizes it to one line
- Clicking on it goes to that message
- We could have an x on the right to dismiss it (?)
- For admins we probably need a menu, with the dismiss option and also to "Unpin for everyone"

![image](https://github.com/user-attachments/assets/e0acc131-5d4a-435d-9413-d538f4933fdf)

**Comment by Chartman123** [comment]
> * floating on the top below the heading

Just my two cents: the floating pinned message hides the contents behind it which triggers me that I want to know whats behind... in the mockup it distracts also from the pinned message if there's some text so close to it. I'd prefer to have a full width bar below the chat info. Text width should of course fit the normal chat width

**Comment by vitspec99** [comment]
Hi
One line "news ticker" just below Chat/Conference name.  On click -- open sidebar with "pinned messages" tab.

**Comment by nickvergessen** [comment]
Rough concept idea:
- [x] Pinned messages are added as a new type to "shared items"
- [x] A conversation property gives a short cut whether the conversation has any pinned conversation and whether the request should be made
- [x] An attendee property is used to "hide" the last pinned message (convo + attendee property have same value => hide)
- [ ] On first hide we should a message that you can still see them in shared items unless unpinned
- [x] There will be system messages (like in other messengers) for pinning at least
- [x] Currently limited to moderators
- [ ] X on the pinned message to hide for "self"
- [ ] Moderators see a second "Unpin" option to revert a pin
- [ ] Pin action is in message … menu
  - [ ] Most likely the chat message needs the info in the meta data as well, so that the action can be changed to "Unpin" in there as well?

**Comment by nickvergessen** [comment]
Question for design @nimishavijay 
- Quotes are single line by now, should also show single line for pinned messages?
- Should we show or hide the avatar / display name / both of the author to show more of the message?

**Comment by SystemKeeper** [comment]
Pin until options:
* 24h
* 7d
* 30d
* Indefinitely
* Custom

**Comment by nimishavijay** [comment]
> Quotes are single line by now, should also show single line for pinned messages?
> Should we show or hide the avatar / display name / both of the author to show more of the message?

Hmm, both interesting ideas. Most other platforms show at least the author of the message and in the name of transparency I say we show who pinned it too.
- [ ] We can show both and keep it 2 lines like in the mockup, but reduce the font size of the author and pinner to font-size-small. 
- [ ] to reduce for wordy it is we can use the pin icon instead of the words "pinned by" so `Marco Ambrosini (📌 Jan C. Borchardt)`
- [ ] I was also thinking about including a date/time when it was pinned to give people an idea of whether it is relevant `Marco Ambrosini (📌 Jan C. Borchardt on July 31st 2025)` what do you think?

**Comment by nickvergessen** [comment]
> Most other platforms show at least the author of the message and in the name of transparency I say we show who pinned it too.

Not sure what you tested, but neither Whatsapp nor Telegram (competing messengers) show anything but the message content, and neither Teams nor Jitsi (competing video calls) have the feature.

That is exactly what made me question this in first place :P

**Comment by nimishavijay** [comment]
I checked Slack (shows the author) and Teams (Teams does have, or at least had the feature because a [help article](https://support.microsoft.com/en-us/office/pin-a-chat-message-to-the-top-of-a-chat-in-microsoft-teams-9c07c454-70bd-4693-a610-21eebca3f136) exists?).

WhatsApp web I do agree that it's very message-forward without any more details. However Telegram saves an entire row for just showing "Pinned message", leaving only one line for the message anyway. So it makes more sense to use that space for something more useful like the author/pinner. We can communicate that it's a pinned message through an icon, what do you think?

**Comment by nickvergessen** [comment]
> Teams (Teams does have, or at least had the feature because a [help article](https://support.microsoft.com/en-us/office/pin-a-chat-message-to-the-top-of-a-chat-in-microsoft-teams-9c07c454-70bd-4693-a610-21eebca3f136) exists?).

Right, I see it in the business one, not in the normal one. But they have 2 different versions:
- In Chat: Author + Message + Time of the message, not the pin
- In Channels: Author + Subject only + Time of the message, not the pin

> We can communicate that it's a pinned message through an icon, what do you think?

I would go with that and then show the Pinner and the pinned_until time in the … menu of the message.
Currently we don't save the pinned_at on the message, if we decide it matters (outside of the system message saying it was pinned) we'd need to add this

**Comment by nimishavijay** [comment]
> Currently we don't save the pinned_at on the message, if we decide it matters (outside of the system message saying it was pinned) we'd need to add this

Saw in the chat that it was added, so it definitely makes sense to to show pinned_at if indefinite, else show pinned_until

**Comment by vitspec99** [comment]
> > Quotes are single line by now, should also show single line for pinned messages?
> > Should we show or hide the avatar / display name / both of the author to show more of the message?
> 
> Hmm, both interesting ideas. Most other platforms show at least the author of the message and in the name of transparency I say we show who pinned it too.
> 
>     * [ ]  We can show both and keep it 2 lines like in the mockup, but reduce the font size of the author and pinner to font-size-small.[ ]  to reduce for wordy it is we can use the pin icon instead of the words "pinned by" so `Marco Ambrosini (📌 Jan C. Borchardt)`[ ]  I was also thinking about including a date/time when it was pinned to give people an idea of whether it is relevant `Marco Ambrosini (📌 Jan C. Borchardt on July 31st 2025)` what do you think?

Two lines "pinned messages block" with different background  just below chat name/info block.  "📌" icon instead of user avatar. On first line: author name and message date like  "Victor Gamov on July 31st 2025:". On second line -- message text like "very long news ticker message."  Any message is displayed for 10 seconds or until the running line ends.  Then next pinned message from another autor appears.

If no pinned messages -- then block is hidden.

**Comment by nickvergessen** [comment]
API is completed including capability and federation support

**Comment by looeee** [comment]
> neither Teams nor Jitsi (competing video calls) have the feature.

Teams does actually have the feature in chats and channels. Here's how it looks in the web/app. 

**Teams chats** (which don't have threads) - clicking the pin banner will bring you to the original message. Only one message can be pinned: 

<img width="1209" height="340" alt="Image" src="https://github.com/user-attachments/assets/198d4937-ad14-4220-a7bc-fd3c0510e322" />


**Teams channels** (which are organized around threads/topics) - you need to open the sidebar, multiple threads can be pinned: 

<img width="1066" height="633" alt="Image" src="https://github.com/user-attachments/assets/bf5d8d00-1b37-4ab2-99c4-33601ace8f37" />

**Comment by SunValleySystems** [comment]
I confirm that this is a necessary feature to convince users to abandon WhatsApp and Teams.
